From ba0a0aefabcf75970c41380aa991c7a12fb23a73 Mon Sep 17 00:00:00 2001
From: Dongli Si <sidongli1997@gmail.com>
Date: Wed, 31 Jul 2024 10:16:36 +0800
Subject: [PATCH] virtio/vhost: Force VIRTIO_NET_F_MAC bit to be enabled

Signed-off-by: Dongli Si <sidongli1997@gmail.com>
---
 virtio/net.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/virtio/net.c b/virtio/net.c
index 492c576..ef2465b 100644
--- a/virtio/net.c
+++ b/virtio/net.c
@@ -514,6 +514,8 @@ static u64 get_host_features(struct kvm *kvm, void *dev)
 		features &= vhost_features;
 	}
 
+	features |= 1UL << VIRTIO_NET_F_MAC;
+
 	return features;
 }
 
@@ -527,7 +529,7 @@ static void virtio_net_start(struct net_dev *ndev)
 			die_perror("TAP device initialized failed because");
 
 		if (ndev->vhost_fd && virtio_vhost_set_features(ndev->vhost_fd,
-								features))
+								features & ~(1UL << VIRTIO_NET_F_MAC)))
 			die_perror("VHOST_SET_FEATURES failed");
 	} else {
 		ndev->info.vnet_hdr_len = virtio_net_hdr_len(ndev);
-- 
2.44.0
